src/compiler/mod.rs

```diff
@@ -114,32 +136,88 @@ impl<'a> Compiler<'a> {
         }
     }
 
+    fn push_to_datatable(&mut self, value: PactType<'a>) -> Result<(), CompileErr> {
+        if self.data_table.len() >= MAX_ENTRIES {
+            Err(CompileErr::DataTableFull)
+        } else {
+            self.data_table.push(value);
+            Ok(())
+        }
+    }
+
     /// Compile an assertion AST node
     fn compile_assertion(&mut self, assertion: &'a ast::Assertion) -> Result<(), CompileErr> {
-        let comparator_op = compile_comparator(&assertion.2)?;
-        self.bytecode.push(comparator_op.into());
-
-        let lhs_load_op = self.compile_subject(&assertion.0)?;
-        match lhs_load_op {
-            OpCode::LD_INPUT(index) | OpCode::LD_USER(index) => {
-                self.bytecode.push(lhs_load_op.into());
-                self.bytecode.push(index);
-            }
-            _ => panic!("unreachable"),
+        let lhs_load = self.compile_subject(&assertion.lhs_subject)?;
+        let (comparator_op, invert) =
+            compile_comparator(&assertion.imperative, &assertion.comparator)?;
+        let rhs_load = self.compile_subject(&assertion.rhs_subject)?;
+
+        // Determine the Load Order
+        let (load, flip) = match lhs_load.load_source {
+            LoadSource::Input => match rhs_load.load_source {
+                LoadSource::Input => (OpLoad::INPUT_VS_INPUT, false),
+                LoadSource::DataTable => (OpLoad::INPUT_VS_USER, false),
+            },
+            LoadSource::DataTable => match rhs_load.load_source {
+                LoadSource::Input => (OpLoad::INPUT_VS_USER, true),
+                _ => return Err(CompileErr::InvalidCompare),
+            },
```

Comment:
style: makes the matches clearer
```suggestion
        let (load, flip) = match (lhs_load.load_source, rhs_load.load_source) {
            (LoadSource::Input, LoadSource::Input)  => (OpLoad::INPUT_VS_INPUT, false),
            (LoadSource::Input, LoadSource::DataTable => (OpLoad::INPUT_VS_USER, false),
            (LoadSource::DataTable, LoadSource::Input)  => (OpLoad::INPUT_VS_USER, true),
            (LoadSource::DataTable, LoadSource::DataTable => return Err(CompileErr::InvalidCompare),
        },
```

Reply:
Nice!